File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/ui/search/AddIndicator.java
package com.battlelancer.seriesguide.ui.search;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.ProgressBar;
import butterknife.BindView;
import butterknife.ButterKnife;
import com.battlelancer.seriesguide.R;

public class AddIndicator extends FrameLayout {

    @BindView(R.id.imageViewAddIndicator) ImageView add;
    @BindView(R.id.imageViewAddIndicatorAdded) ImageView added;
    @BindView(R.id.progressBarAddIndicator) ProgressBar progressBar;

    public AddIndicator(Context context, AttributeSet attrs) {
        super(context, attrs);

        LayoutInflater.from(context).inflate(R.layout.layout_add_indicator, this);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        ButterKnife.bind(this);

        added.setVisibility(GONE);
        progressBar.setVisibility(GONE);
    }

    public void setContentDescriptionAdded(CharSequence contentDescription) {
        added.setContentDescription(contentDescription);
    }

    public void setOnAddClickListener(OnClickListener onClickListener) {
        add.setOnClickListener(onClickListener);
    }

    public void setState(int state) {
        if (state == SearchResult.STATE_ADD) {
            add.setVisibility(VISIBLE);
            progressBar.setVisibility(GONE);
            added.setVisibility(GONE);
        } else if (state == SearchResult.STATE_ADDING) {
            add.setVisibility(GONE);
            progressBar.setVisibility(VISIBLE);
            added.setVisibility(GONE);
        } else if (state == SearchResult.STATE_ADDED) {
            add.setVisibility(GONE);
            progressBar.setVisibility(GONE);
            added.setVisibility(VISIBLE);
        }
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/shows/search/discover/AddIndicator.kt
// Copyright 2023 Uwe Trottmann
// SPDX-License-Identifier: Apache-2.0

package com.battlelancer.seriesguide.shows.search.discover

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.FrameLayout
import androidx.appcompat.widget.TooltipCompat
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.databinding.LayoutAddIndicatorBinding

/**
 * A three state visual indicator with states add, adding and added from [SearchResult].
 *
 * Make sure to call [setNameOfAssociatedItem] to provide content description and tooltips for all
 * states.
 */
class AddIndicator(context: Context, attrs: AttributeSet?) : FrameLayout(context, attrs) {

    private val binding = LayoutAddIndicatorBinding.inflate(LayoutInflater.from(context), this)

    override fun onFinishInflate() {
        super.onFinishInflate()
        binding.imageViewAddIndicator.also {
            TooltipCompat.setTooltipText(it, it.contentDescription)
        }
        binding.imageViewAddIndicatorAdded.visibility = GONE
        binding.progressBarAddIndicator.visibility = GONE
    }

    /**
     * Sets content descriptions and tooltips of added and adding states.
     */
    fun setNameOfAssociatedItem(name: String) {
        binding.apply {
            imageViewAddIndicatorAdded.also {
                it.contentDescription = context.getString(R.string.add_already_exists, name)
                TooltipCompat.setTooltipText(it, it.contentDescription)
            }
            progressBarAddIndicator.also {
                it.contentDescription = context.getString(R.string.add_started, name)
                TooltipCompat.setTooltipText(it, it.contentDescription)
            }
        }
    }

    fun setOnAddClickListener(onClickListener: OnClickListener?) {
        binding.imageViewAddIndicator.setOnClickListener(onClickListener)
    }

    fun setState(state: Int) {
        when (state) {
            SearchResult.STATE_ADD -> {
                binding.imageViewAddIndicator.visibility = VISIBLE
                binding.progressBarAddIndicator.visibility = GONE
                binding.imageViewAddIndicatorAdded.visibility = GONE
            }

            SearchResult.STATE_ADDING -> {
                binding.imageViewAddIndicator.visibility = GONE
                binding.progressBarAddIndicator.visibility = VISIBLE
                binding.imageViewAddIndicatorAdded.visibility = GONE
            }

            SearchResult.STATE_ADDED -> {
                binding.imageViewAddIndicator.visibility = GONE
                binding.progressBarAddIndicator.visibility = GONE
                binding.imageViewAddIndicatorAdded.visibility = VISIBLE
            }
        }
    }
}